library/core/src/backtrace.rs

```diff
@@ -0,0 +1,160 @@
+//! hi
+#![unstable(feature = "core_backtrace", issue = "74465")]
+use crate::fmt;
+
+// perma(?)-unstable
+#[unstable(feature = "core_backtrace", issue = "74465")]
+///
+pub trait RawBacktraceImpl: fmt::Debug + fmt::Display + 'static {
+    ///
+    unsafe fn drop_and_free(self: *mut Self);
+}
+
+/// The current status of a backtrace, indicating whether it was captured or
+/// whether it is empty for some other reason.
+#[non_exhaustive]
+#[derive(Debug, PartialEq, Eq)]
+#[unstable(feature = "core_backtrace", issue = "74465")]
+pub enum BacktraceStatus {
+    /// Capturing a backtrace is not supported, likely because it's not
+    /// implemented for the current platform.
+    Unsupported,
+    /// Capturing a backtrace has been disabled through either the
+    /// `RUST_LIB_BACKTRACE` or `RUST_BACKTRACE` environment variables.
+    Disabled,
+    /// A backtrace has been captured and the `Backtrace` should print
+    /// reasonable information when rendered.
+    Captured,
+}
+
+///
+#[unstable(feature = "core_backtrace", issue = "74465")]
+#[derive(Debug)]
+pub enum Inner {
+    ///
+    Unsupported,
+    ///
+    Disabled,
+    ///
+    Captured(*mut dyn RawBacktraceImpl),
+}
+
+#[unstable(feature = "core_backtrace", issue = "74465")]
+///
+pub struct Backtrace {
+    inner: Inner,
+}
+
+extern "Rust" {
+    // #[lang_item = "backtrace_capture"]
+    fn backtrace_capture() -> Inner;
+    // #[lang_item = "backtrace_enabled"]
+    fn backtrace_enabled() -> bool;
+}
+
+impl Backtrace {
+    /// Capture a stack backtrace of the current thread.
+    ///
+    /// This function will capture a stack backtrace of the current OS thread of
+    /// execution, returning a `Backtrace` type which can be later used to print
+    /// the entire stack trace or render it to a string.
+    ///
+    /// This function will be a noop if the `RUST_BACKTRACE` or
+    /// `RUST_LIB_BACKTRACE` backtrace variables are both not set. If either
+    /// environment variable is set and enabled then this function will actually
+    /// capture a backtrace. Capturing a backtrace can be both memory intensive
+    /// and slow, so these environment variables allow liberally using
+    /// `Backtrace::capture` and only incurring a slowdown when the environment
+    /// variables are set.
+    ///
+    /// To forcibly capture a backtrace regardless of environment variables, use
+    /// the `Backtrace::force_capture` function.
+    #[inline(never)] // want to make sure there's a frame here to remove
+    #[unstable(feature = "core_backtrace", issue = "74465")]
+    pub fn capture() -> Self {
```

Comment:
One design that was floated in our Libs meeting was to define `Backtrace::capture` and co in `std`, so there would be no way (yet) to capture a `Backtrace` in `core`, even though we could refer to one there. That could remove the need for figuring out hooks right now. Supposedly adding impl blocks in `std` to types defined in `core` is something we can already do. An alternative option could be a free `backtrace::capture` function.

So this would look something like `RawWaker::new` where it accepts a `*mut dyn RawBacktrace`.

Reply:
I am in favor of this idea. As added bonus it would probably (unstably) allow arbitrary backtraces to be constructed for for example language level backtraces in interpreters or jit engines. Or allow mixing of language level and native frames.

Reply:
I'm lacking context but what's left to figure out, does this code not work as written?

As I mentioned in the other thread, the compiler [has a usecase](https://github.com/rust-lang/rust/issues/79153) for capturing a backtrace in core (in signal handlers), so it would be nice if there were some unstable path we could call to do it.

Reply:
@tmandry I think the missing piece here is a hook that makes `Backtrace::capture` do something meaningful when `std` is linked in. We could avoid that need by defining the `capture` function itself in `std` either on the `Backtrace` type or as a free function.

Does the compiler usecase need to capture a `Backtrace` type? Or just print one? We would still have an (unstable) way to build a `Backtrace` in `core`, it just wouldn't be on a path to stabilization yet.

Reply:
We just need to print one, creating an actual `Backtrace` value is not in scope. But printing is still a lot of logic I don't want to duplicate.

Reply:
Yeh, it's a lot to keep in sync already between `panic!` and `Backtrace::capture` output too, so was wondering whether some non-allocating equivalent to the `backtrace` crate's frames iterator for printing could be worthwhile for that.

Reply:
I started the PR over again to try to work in the suggestions here but im currently stuck at the "implement methods on a foreign type" step. It seems we need to use a lang item to do this but @eddyb is worried that using that for anything other than a builtin type could cause issues with coherence.

Reply:
It doesn't need to be a lang item. It could be a method on an extension trait defined inside libstd. `core::backtrace::Backtrace` then only has to have a public unstable constructor that can be used within the extension trait to create it.

Reply:
> It doesn't need to be a lang item. It could be a method on an extension trait defined inside libstd. core::backtrace::Backtrace then only has to have a public unstable constructor that can be used within the extension trait to create it.

Follow up questions:

* would users then need to bring this extension trait into scope in order to use any methods on backtrace? Would we want to add this trait to the prelude?
* Is the extension trait unstable? How do we add new methods to backtrace in the future? Is it sealed?

Reply:
The trait should be stable but sealed. Whether or not to add it to the prelude I don't know.

Reply:
Maybe relevant: I recently discovered that `use FooTrait as _;` works inside the prelude (or some other module with a wildcard import), pulling it in scope without polluting the namespace.

Reply:
Question: Are we confident that people won't want to eventually trigger backtrace capture in `no_std` code bases (libraries) where the backtrace uses `std` machinery if that library is linked into an application that includes std?

I realized that the recent suggestions for ways we could accomplish the core backtrace move all seem to discard this feature, but I'm not confident that this isn't a useful feature or that going down these alternate paths might make it harder to add this feature in the future. My instinct is that I still want to lean towards the allocator lang item inspired design rather than an impl block lang item design, since either way they seem to be using the same tricks to break the coherence rules.